src/lib.rs

```diff
@@ -225,74 +225,89 @@ impl<Message> MessageFilter<Message> where Message: PartialOrd + Ord + Clone + :
+        assert!((0..size).all(|index| msg_filter.check(&index)));
+
+        // Add further messages - all should be added, each time pushing out the oldest message.
+        for i in size..1000 {
             msg_filter.add(i);
             assert_eq!(msg_filter.len(), size);
-        }
-
-        for _ in (0..1000).rev() {
-            assert!(msg_filter.check(&(1000 - 1)));
+            assert!(msg_filter.check(&i));
+            assert!(msg_filter.check(&(i - 1)));
+            assert!(msg_filter.check(&(i - size + 1)));
+            assert!(!msg_filter.check(&(i - size)));
         }
     }
 
     #[test]
     fn time_only() {
-        let time_to_live = ::time::Duration::milliseconds(100);
+        use ::rand::Rng;
+        let time_to_live = ::time::Duration::milliseconds(::rand::thread_rng().gen_range(50, 150));
         let mut msg_filter = super::MessageFilter::<usize>::with_expiry_duration(time_to_live);
+        assert_eq!(time_to_live, msg_filter.time_to_live);
+        assert_eq!(::std::usize::MAX, msg_filter.capacity);
 
+        // Add 10 messages - all should be added.
         for i in 0..10 {
-            assert_eq!(msg_filter.len(), i);
             msg_filter.add(i);
-            assert_eq!(msg_filter.len(), i + 1);
+            assert!(msg_filter.check(&i));
         }
+        assert_eq!(msg_filter.len(), 10);
 
-        ::std::thread::sleep_ms(100);
-        msg_filter.add(11);
+        // Allow the added messages time to expire.
+        ::std::thread::sleep_ms(time_to_live.num_milliseconds() as u32 + 10);
 
+        // Add a new message which should cause the expired values to be removed.
+        msg_filter.add(11);
+        assert!(msg_filter.check(&11));
         assert_eq!(msg_filter.len(), 1);
 
+        // Check we can add the initial messages again.
         for i in 0..10 {
             assert_eq!(msg_filter.len(), i + 1);
             msg_filter.add(i);
+            assert!(msg_filter.check(&i));
             assert_eq!(msg_filter.len(), i + 2);
         }
     }
 
     #[test]
     fn time_and_size() {
-        let size = 10usize;
-        let time_to_live = ::time::Duration::milliseconds(100);
+        use ::rand::Rng;
+        let size = ::rand::random::<u8>() as usize;
+        let time_to_live = ::time::Duration::milliseconds(::rand::thread_rng().gen_range(50, 150));
         let mut msg_filter =
             super::MessageFilter::<usize>::with_expiry_duration_and_capacity(time_to_live, size);
+        assert_eq!(time_to_live, msg_filter.time_to_live);
+        assert_eq!(size, msg_filter.capacity);
 
         for i in 0..1000 {
+            // Check `size` has not been exceeded.
             if i < size {
                 assert_eq!(msg_filter.len(), i);
+            } else {
+                assert_eq!(msg_filter.len(), size);
             }
 
+            // Add a new message and check that it has been added successfully.
             msg_filter.add(i);
+            assert!(msg_filter.check(&i));
```

Comment:
Good point.  I think we should actually not allow construction of MessageFilter if cacpacity or TTL are 0, but that's a breaking API change (we'd probably want to return a Result in the c'tors).

I'll fix the test to not allow size of zero for now and add an issue.